doc(v3): add v3 roadmap and design document (#1583)
zarr-developers/zarr-python: v3-roadmap-and-design.md

> # Zarr Python Roadmap

Author: This is formatted w/ markdown. If folks think this doc would fit better as a page in the docs, we can convert it to RST. cc @MSanKeys963

During the development of the V3 Specification, a [prototype implementation](https://github.com/zarr-developers/zarr-python/pull/898) was added to the Zarr-Python library. Since that implementation, the V3 spec evolved in significant ways and as a result, the Zarr-Python library is now out of sync with the approved spec. Downstream libraries (e.g. [Xarray](https://github.com/pydata/xarray)) have added support for this implementation and will need to migrate to the accepted spec when its available in Zarr-Python.

## Goals

- Provide a complete implementation of Zarr V3 through the Zarr-Python API
- Align the Zarr-Python array API with the [array API Standard](https://data-apis.org/array-api/latest/)
- Clear the way for exciting extensions / ZEPs (i.e. [sharding](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/sharding-indexed/v1.0.html), [variable chunking](https://zarr.dev/zeps/draft/ZEP0003.html), etc.)
- Provide a developer API that can be used to implement and register V3 extensions
- Improve the performance of Zarr-Python by streamlining the interface between the Store layer and higher level APIs (e.g. Groups and Arrays)
- Clean up the internal and user facing APIs 

> - Improve code quality and robustness (e.g. achieve 100% type hint coverage)

Reviewer: :+1: for prioritization of the goals (is the list already ordered?)

> - Provide a complete implementation of Zarr V3 through the Zarr-Python API
> - Align the Zarr-Python array API with the [array API Standard](https://data-apis.org/array-api/latest/)
> - Clear the way for exciting extensions / ZEPs (i.e. [sharding](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/sharding-indexed/v1.0.html), [variable chunking](https://zarr.dev/zeps/draft/ZEP0003.html), etc.)
> - Provide a developer API that can be used to implement and register V3 extensions
> - Improve the performance of Zarr-Python by streamlining the interface between the Store layer and higher level APIs (e.g. Groups and Arrays)
> - Clean up the internal and user facing APIs 
> - Improve code quality and robustness (e.g. achieve 100% type hint coverage)

Author: I had intentionally not ordered but here is my own view on priorities here.

```suggestion
- Provide a complete implementation of Zarr V3 through the Zarr-Python API
- Clear the way for exciting extensions / ZEPs (i.e. [sharding](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/sharding-indexed/v1.0.html), [variable chunking](https://zarr.dev/zeps/draft/ZEP0003.html), etc.)
- Provide a developer API that can be used to implement and register V3 extensions
- Improve the performance of Zarr-Python by streamlining the interface between the Store layer and higher level APIs (e.g. Groups and Arrays)
- Clean up the internal and user facing APIs 
- Improve code quality and robustness (e.g. achieve 100% type hint coverage)
- Align the Zarr-Python array API with the [array API Standard](https://data-apis.org/array-api/latest/)
```

## Examples of what 3.0 will enable?
1. Reading and writing V3 spec-compliant groups and arrays

> 2. V3 extensions including sharding and variable chunking.

Reviewer: It's a good question but might be outside the scope of this document (since it's unsettled anyway).

> 2. V3 extensions including sharding and variable chunking.

Author: We've been proceeding assuming that some of the extensions that motivated V3 (sharding, variable chunk grids) would eventually find support here. But we aren't requiring full support for these extensions for the initial release.

> 2. V3 extensions including sharding and variable chunking.

Reviewer 2: Fwiw, the sharding codec is already in the v3 branch. So, it could ship in the initial release.

3. Consolidation of writes (e.g. array metadata + attributes concurrently)
4. Improved concurrency when creating/reading/writing to stores (imagine a `create_hierarchy(zarr_objects)` function).
5. User-developed extensions (e.g. storage-transformers) can be registered with Zarr-Python at runtime

## Non-goals (of this document)

- Implementation of any Zarr V3 extensions
- Major revisions to the Zarr V3 spec

## Requirements

1. Read and write spec compliant V2 and V3 data
2. Limit unnecessary traffic to/from the store
3. Cleanly define the Array/Group/Store abstractions
4. Cleanly define how V2 will be supported going forward
5. Provide a clear roadmap to help users upgrade to 3.0
6. Developer tools / hooks for registering extensions

## Design

### Async API

> Zarr-Python is an IO library. As such, supporting concurrent action against the storage layer is critical to achieving acceptable performance. The Zarr-Python 2 was not designed with asynchronous computation in mind and as a result has struggled to effectively leverage the benefits of concurrency. At one point, `getitems` and `setitems` support was added to the Zarr store model but that is not leveraged throughout the API. 

Author: ```suggestion
Zarr-Python is an IO library. As such, supporting concurrent action against the storage layer is critical to achieving acceptable performance. The Zarr-Python 2 was not designed with asynchronous computation in mind and as a result has struggled to effectively leverage the benefits of concurrency. At one point, `getitems` and `setitems` support was added to the Zarr store model but that is only used for operating on a set of chunks in a single variable. 
```

With Zarr-Python 3.0, we have the opportunity to revisit this design. The proposal here is as follows:

1. The `Store` interface will be entirely async. 
2. On top of the async `Store` interface, we will provide an  `AsyncArray` and `AsyncGroup` interface.

> 3. Finally, the primary user facing API will be synchronous `Array` and `Group` classes that wrap the async equivalents.

Reviewer 3: > We probably also need some concept of "batch" and/or "queue" to allow files to accumulate or stream to the async IO.

I think tensorstore achieves something like this via [transactions](https://google.github.io/tensorstore/python/api/tensorstore.Transaction.html), it might be worth copying that API

> 3. Finally, the primary user facing API will be synchronous `Array` and `Group` classes that wrap the async equivalents.

Author: We have been discussing something similar that could exist at (or near) the store layer. 

```python

store = MyStore()

with store.transaction() as ts:
    array_foo.create(..., store=ts)
    array_bar.create(..., store=ts)
```

This is going to necessitate some thought around the state of the store before the transaction but would enable matching multiple writes in the form of a write-cache.

> 3. Finally, the primary user facing API will be synchronous `Array` and `Group` classes that wrap the async equivalents.

Reviewer 2: Wouldn't it be possible for users to use the AsyncArray and AsyncGroup directly, if concurrency is important to them?

**Examples**

- **Store**
    
    ```python
    class Store:
        ...
        async def get(self, key: str) -> bytes:
            ...

>         async def get_partial_values(self, key_ranges: List[Tuple[str, int]]) -> bytes:

Reviewer 4: This is mostly to make sure im understanding as im spinning up, but this method would need the range _length_ also in the type for `key_ranges`? eg
```
Range = Tuple[int, Optional[int]]
...

key_ranges: List[Tuple[str, Range]]
```

Author: Good catch @aldenks!

3. Will Zarr help manage the async contexts encouraged by some libraries (e.g. [AioBotoCore](https://aiobotocore.readthedocs.io/en/latest/tutorial.html#using-botocore))?
  a. Many async IO libraries require entering an async context before interacting with the API. We expect some experimentation to be needed here but the initial design will follow something close to what fsspec does ([example in s3fs](https://github.com/fsspec/s3fs/blob/949442693ec940b35cda3420c17a864fbe426567/s3fs/core.py#L527)).
4. Why not provide a synchronous Store interface?
  a. We could but this design is simpler. It would mean supporting it in the `AsyncGroup` and `AsyncArray` classes which, may be more trouble than its worth.

>   b. One option would be to provide a `SyncStore` interface and an `AsyncStoreWrapper` that wraps each method in a `loop.run_in_executor`. This would help avoid long-running blocking tasks in the main loop.

Author: Same here. I suspect someone will want to do this eventually but there is a clear path toward wrapping a sync api so I'm not too worried about it.

### Store API

The `Store` API is specified directly in the V3 specification. All V3 stores should implement this abstract API, omitting Write and List support as needed. As described above, all stores will be expected to expose the required methods as async methods.

**Example**

```python
class ReadWriteStore:
		...
    async def get(self, key: str) -> bytes:
        ...

    async def get_partial_values(self, key_ranges: List[Tuple[str, int]]) -> bytes:
        ...
		

>     async def set(self, key: str, value: bytes) -> None:

Reviewer 2: Input arguments should probably be something like [BytesLike](https://github.com/scalableminds/zarrita/blob/66ea5d65c89e42c06a46dfa1f2a7c761aab8615e/zarrita/common.py#L27). I think an optional output argument could be useful, too. 

I want to mention that the codec pipeline potentially also makes several copies of the data. There is certainly room for optimization on that front.

        ...  # required for writable stores

    async def set_partial_values(self, key_start_values: List[Tuple[str, int, bytes]]) -> None:
        ...  # required for writable stores

    async def list(self) -> List[str]:
        ...  # required for listable stores

    async def list_prefix(self, prefix: str) -> List[str]:
        ...  # required for listable stores

    async def list_dir(self, prefix: str) -> List[str]:
        ...  # required for listable stores

    # additional (optional methods)
    async def getsize(self, prefix: str) -> int:
        ...
        
    async def rename(self, src: str, dest: str) -> None
        ...
        
```

> Recognizing that there are many Zarr applications today that rely on the `MutableMapping` interface supported by Zarr-Python 2, a wrapper store will be developed to allow existing stores to plug directly into this API.

Reviewer 2: I think it would be most elegant if users could supply MutableMapping stores through the high-level API that are then auto-wrapped.

| | `ndim` | `T` | |
| | `shape` | | |
| | `size` | | |
| Methods | `__getitem__` | `__array_namespace__` | `to_device` |
| | `__setitem__` | `__abs__` | `__bool__` |
| | `__eq__` | `__add__` | `__complex__` |
| | `__bool__` | `__and__` | `__dlpack__` |
| | | `__floordiv__` | `__dlpack_device__` |
| | | `__ge__` | `__float__` |
| | | `__gt__` | `__index__` |
| | | `__invert__` | `__int__` |
| | | `__le__` | |
| | | `__lshift__` | |
| | | `__lt__` | |
| | | `__matmul__` | |
| | | `__mod__` | |
| | | `__mul__` | |
| | | `__ne__` | |
| | | `__neg__` | |
| | | `__or__` | |
| | | `__pos__` | |
| | | `__pow__` | |
| | | `__rshift__` | |
| | | `__sub__` | |
| | | `__truediv__` | |
| | | `__xor__` | |
| Creation functions (`zarr.creation`) | `zeros` | | `arange` |
| | `zeros_like` | | `asarray` |
| | `ones` | | `eye` |
| | `ones_like` | | `from_dlpack` |
| | `full` | | `linspace` |
| | `full_like` | | `meshgrid` |
| | `empty` | | `tril` |
| | `empty_like` | | `triu` |

In addition to the core array API defined above, the Array class should have the following Zarr specific properties:

- `.metadata` (see Metadata Interface below)
- `.attrs` - (pull from metadata object)

> - `.info` - (pull from existing property)

Author: Agreed that listing the store is not a great default behavior. I'll add a note here.

| `full` | `full` | `full` | N/A |
| `array` | `create_array` | `create_array` | N/A |
| `empty_like` | `empty_like` | `empty_like` | N/A |
| `zeros_like` | `zeros_like` | `zeros_like` | N/A |
| `ones_like` | `ones_like` | `ones_like` | N/A |
| `full_like` | `full_like` | `full_like` | N/A |
| `move` | `move` | `move` | `move` |

**`zarr.h5compat.Group`**

Zarr-Python 2.* made an attempt to align its API with that of [h5py](https://docs.h5py.org/en/stable/index.html). With 3.0, we will relax this alignment in favor of providing an explicit compatibility module (`zarr.h5py_compat`). This module will expose the `Group` and `Dataset` APIs that map to Zarr-Python’s `Group` and `Array` objects. 

### Creation API

Zarr-Python 2.* bundles together the creation and serialization of Zarr objects. Zarr-Python 3.* will make it possible to create objects in memory separate from serializing them. This will specifically enable writing hierarchies of Zarr objects in a single batch step. For example:

```python

arr1 = Array(shape=(10, 10), path="foo/bar", dtype="i4", store=store)
arr2 = Array(shape=(10, 10), path="foo/spam", dtype="f8", store=store)

arr1.save()
arr2.save()

# or equivalently

> zarr.save_many([arr1 ,arr2])

Reviewer 5: This is a big point! So here is the batching that I mentioned before. It should be clarified, what the state of `store` is before save/save_many is called: is it usable, does it have some ephemeral storage (including data?) ?

Author: Good questions @martindurant - I've left a note that indicates this part of the doc is in need of some additional design effort.

```

### Plugin API

Zarr V3 was designed to be extensible at multiple layers. Zarr-Python will support these extensions through a combination of [Abstract Base Classes](https://docs.python.org/3/library/abc.html) (ABCs) and [Entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/). 

**ABCs**

Zarr V3 will expose Abstract base classes for the following objects:

- `Store`, `ReadStore`, `ReadWriteStore`, `ReadListStore`, and `ReadWriteListStore`
- `BaseArray`, `SynchronousArray`, and `AsynchronousArray`
- `BaseGroup`, `SynchronousGroup`, and `AsynchronousGroup`

> - `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`

Author: @normanrz mentioned the idea about writing a set of codec that would wrap `numcodecs` existing compressors/filters/etc. That would likely be the cleanest way to continue using numcodecs going forward. With the entrypoint api for codecs, the wrapper class could actually live in numcodecs and simply be registered as an entrypoint that follows the protocol. See #1588 for a start on this api.

> - `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`

Author: This also came up in today's refactor WG call. @normanrz has given this more thought than I have. Perhaps he can share a bit more on what you think is needed here?

> - `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`

Reviewer 2: The Zarr 3 spec is pretty explicit about what codecs are in spec to ensure compatibility among implementations. That set of codecs is now implemented directly in zarr-python (with use of numcodecs).

However, it is desirable to experiment with other codecs in zarr-python and make use of numcodecs. I think there are 2 ways of doing this:
1. We could make generic adapters for numcodecs to fit the zarr codecs API (ie. ArrayArrayCodec, ArrayBytesCodec, BytesBytesCodec). That would also be useful to code v2 compatibility.
2. We could also introduce a new API in numcodecs that presents all codecs using the Zarr codec API. Not sure what the appetite for such a change on the numcodecs side is.

I am a bit concerned about the codec metadata that gets written to the jsons because most codecs in numcodecs are not in the Zarr v3 spec (or with different metadata). Maybe we should actually mark them as numcodecs codecs:
```
"codecs": [{"name": "numcodecs", "configuration": {"id":"blosc", "cname": "zstd", ...}}]
```

> - `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`

Reviewer 6: Thanks for raising this, @jeromekelleher, and the explanation here, @jhamman and @normanrz.

I was wondering what would be the ideal way to add a new codec (if and when we plan to add any) in Zarr-Python V3.0?
And what would be the best way forward out of the two options if we consider the possibility of extending the existing codecs in the V3 specification?

Also, it'd be great if we could get @jakirkham views on this.

> - `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`

Reviewer 2: To add new codecs to zarr-python, developers can subclass the base classes (e.g. `ArrayBytesCodec`, `ArrayArrayCodec`, `BytesBytesCodec`) and register the codec using the `[zarr.codecs]` entry point in their pyproject.toml or programmatically call [`zarr.v3.codecs.registry.register_codec`](https://github.com/zarr-developers/zarr-python/blob/v3/zarr/v3/codecs/registry.py#L31).

**Entrypoints**

Lots more thinking here but the idea here is to provide entrypoints for `data type`, `chunk grid`, `chunk key encoding`, `codecs`,  `storage_transformers` and `stores`.  These might look something like:

```
entry_points="""
    [zarr.codecs]
    blosc_codec=codec_plugin:make_blosc_codec
    zlib_codec=codec_plugin:make_zlib_codec
"""
```

### Python type hints and static analysis

Target 100% Mypy coverage in 3.0 source. 

### Observability

A persistent problem in Zarr-Python is diagnosing problems that span many parts of the stack. To address this in 3.0, we will add a basic logging framework that can be used to debug behavior at various levels of the stack. We propose to add the separate loggers for the following namespaces:

- `array`
- `group`
- `store`
- `codec`

These should be documented such that users know how to activate them and developers know how to use them when developing extensions.

### Dependencies

Today, Zarr-Python has the following required dependencies:

```python
dependencies = [
    'asciitree',
    'numpy>=1.20,!=1.21.0',
    'fasteners',
    'numcodecs>=0.10.0',
]
```

What other dependencies should be considered?

> 1. Attrs - Zarrita makes extensive use of the Attrs library

Reviewer 3: I'm sympathetic to this perspective @jeromekelleher. I could imagine switching from attrs to dataclasses (or even just vanilla python classes :) ) once the API changes have cooled down.

> 1. Attrs - Zarrita makes extensive use of the Attrs library

Reviewer 2: I agree that we should remove attrs. However, for development right now we can probably move faster when keeping it in right now and later do one concerted effort to replace it.

> 2. Fsspec - Zarrita has a hard dependency on Fsspec. This could be easily relaxed though.

Reviewer 5: (noting that fsspec has no dependencies of its own, but many of its implementations do)

> ## Breaking changes relative to Zarr-Python 2.*

Reviewer 5: Part of the problem is not having a clear definition of which part of v2 is considered to outward-facing API. I would suggest it's Group and Array *only* (plus a few convenience functions like open_*).

1. H5py compat moved to a stand alone module?
2. `Group.__getitem__` support moved to `Group.members.__getitem__`?

> 3. Others? 

Reviewer 7: This is really important point @jeromekelleher! The doc says there will be breaking changes to APIs, but doesn't specify what those changes will be. In order for the community to get on board, we should be super clear about this. My hope is that breaking changes are very minimal, and whatever breaking changes there are should be justified with a clearly argued rationale.

> 3. Others? 

Author: To be honest, we're still sorting out the feasibility of maintaining API compatibility at the Array level. @d-v-b has been leading this so I'll let him share his thoughts.

> 3. Others? 

Reviewer 3: There are a couple fundamental challenges for preserving the `zarr-python` v2 api after we fully integrate v3: The `Array` class in `zarr-python` exposes all the attributes defined for zarr version 2 arrays. But zarr version 3 arrays have different attributes, so in `zarr-python` version 3 we have a few options for dealing with this: 
1. Uses the current `zarr-python` `Array` class to wrap v3 arrays and make them look like v2 arrays. this maximizes backwards compatibility, but would be a usability nightmare and confusing for people who want to work primarily with v3 arrays. 
2. Create a new `Array` class in `zarr-python` that exposes zarr version 3 array attributes, much like `zarr-python` v2 did for zarr version 2 arrays, and then use this class to wrap zarr version 2 arrays. This would also be unpleasant, because zarr version 3 arrays don't have a `dtype` property (instead, that attribute is called `data_type`), and so hewing too close the zarr version 3 array spec for `zarr-python` would result in an impedance mismatch with the wider pydata ecosystem.
3. elevate the `Array` class to a level of abstraction above zarr version 2 and version 3 array specifications, and make that `Array` class capable of using zarr version 2 or 3 arrays as load-bearing members. This lets us optimize the `Array` class in zarr-python for a specific goal: a performant, familiar API for people coming from other pydata tools, while smoothing over the differences between v2 and v3 arrays. I think this is the best option, but it does involve substantial breaking changes to the `zarr-python` array API.

The only option here that doesn't break the `zarr-python` array API is 1, but it's a terrible option for people who want to get the most out of zarr version 3. The fact is, zarr v3 is a breaking change w.r.t. zarr v2, and I would suggest that people who do not want their code to break should continue using zarr v2 and pin `zarr-python` to version 2.

And this doesn't touch performance at all. I think we will need to change a lot of APIs to get good performance in zarr-python version 3.

## Goals

- Provide a complete implementation of Zarr V3 through the Zarr-Python API
- Clear the way for exciting extensions / ZEPs (i.e. [sharding](https://zarr-specs.readthedocs.io/en/latest/v3/codecs/sharding-indexed/v1.0.html), [variable chunking](https://zarr.dev/zeps/draft/ZEP0003.html), etc.)
- Provide a developer API that can be used to implement and register V3 extensions
- Improve the performance of Zarr-Python by streamlining the interface between the Store layer and higher level APIs (e.g. Groups and Arrays)
- Clean up the internal and user facing APIs 
- Improve code quality and robustness (e.g. achieve 100% type hint coverage)
- Align the Zarr-Python array API with the [array API Standard](https://data-apis.org/array-api/latest/)

## Examples of what 3.0 will enable?
1. Reading and writing V3 spec-compliant groups and arrays
2. V3 extensions including sharding and variable chunking.
3. Improved performance by leveraging concurrency when creating/reading/writing to stores (imagine a `create_hierarchy(zarr_objects)` function).
4. User-developed extensions (e.g. storage-transformers) can be registered with Zarr-Python at runtime

## Non-goals (of this document)

> - Implementation of any Zarr V3 extensions

Reviewer 2: ```suggestion
- Implementation of any unaccepted Zarr V3 extensions
```

    
        def create_group(self, path: str, **kwargs) -> Group:
            return sync(self._async_group.create_group(path, **kwargs))
    ```
**Internal Synchronization API**

With the `Store` and core `AsyncArray`/ `AsyncGroup` classes being predominantly async, Zarr-Python will need an internal API to provide a synchronous API. The proposal here is to use the approach in [fsspec](https://github.com/fsspec/filesystem_spec/blob/master/fsspec/asyn.py) to provide a high-level `sync` function that takes an `awaitable` and runs it in its managed IO Loop / thread. 

**FAQ**
1. Why two levels of Arrays/groups?
  a. First, this is an intentional decision and departure from the current Zarrita implementation
  b. The idea is that users rarely want to mix interfaces. Either they are working within an async context (currently quite rare) or they are in a typical synchronous context. 
  c. Splitting the two will allow us to clearly define behavior on the `AsyncObj` and simply wrap it in the `SyncObj`.
2. What if a store is only has a synchronous backend?
  a. First off, this is expected to be a fairly rare occurrence. Most storage backends have async interfaces.

>   b. But in the even a storage backend doesn’t have a async interface, there is nothing wrong with putting synchronous code in `async` methods. There are approaches to enabling concurrent action through wrappers like AsyncIO's `loop.run_in_executor` ([ref 1](https://stackoverflow.com/questions/38865050/is-await-in-python3-cooperative-multitasking ), [ref 2](https://stackoverflow.com/a/43263397/732596), [ref 3](https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-5.html), [ref 4](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).  

Reviewer 6: ```suggestion
  b. But in the event a storage backend doesn’t have a async interface, there is nothing wrong with putting synchronous code in `async` methods. There are approaches to enabling concurrent action through wrappers like AsyncIO's `loop.run_in_executor` ([ref 1](https://stackoverflow.com/questions/38865050/is-await-in-python3-cooperative-multitasking ), [ref 2](https://stackoverflow.com/a/43263397/732596), [ref 3](https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-5.html), [ref 4](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor).  
```
